vimiv/imutils/exif.py

```diff
@@ -11,138 +11,267 @@
+                                    piexif.TYPES.SLong,
+                                    piexif.TYPES.Float,
+                                    piexif.TYPES.DFloat,
+                                ):  # integer and float
+                                    exif[keyname] = (keyname, val)
+                                elif keytype in (
+                                    piexif.TYPES.Ascii,
+                                    piexif.TYPES.Undefined,
+                                ):  # byte encoded
+                                    exif[keyname] = (keyname, val.decode())
+                                elif keytype in (
+                                    piexif.TYPES.Rational,
+                                    piexif.TYPES.SRational,
+                                ):  # (int, int) <=> numerator, denominator
+                                    exif[keyname] = (keyname, f"{val[0]}/{val[1]}")
+
+                    except (piexif.InvalidImageDataError, KeyError):
+                        return None
+
+                    return exif
+
+                def copy_exif(self, dest: str, reset_orientation: bool = True) -> None:
+                    """Copy exif information from current image to dest.
+
+                    Args:
+                        dest: Path to write the exif information to.
+                        reset_orientation: If true, reset the exif orientation tag to
+                            normal.
+                    """
+
+                    try:
+                        if reset_orientation:
+                            with contextlib.suppress(KeyError):
+                                self._metadata["0th"][
+                                    piexif.ImageIFD.Orientation
+                                ] = ExifOrientation.Normal
+                        exif_bytes = piexif.dump(self._metadata)
+                        piexif.insert(exif_bytes, dest)
+                        _logger.debug("Successfully wrote exif data for '%s'", dest)
+                    except piexif.InvalidImageDataError:  # File is not a jpg
+                        _logger.debug(
+                            "File format for '%s' does not support exif", dest
+                        )
+                    except ValueError:
+                        _logger.debug("No exif data in '%s'", dest)
+
+                def exif_date_time(self) -> str:
+                    """Exif creation date and time of filename."""
+
+                    with contextlib.suppress(
+                        piexif.InvalidImageDataError, FileNotFoundError, KeyError
+                    ):
+                        return self._metadata["0th"][piexif.ImageIFD.DateTime].decode()
+                    return ""
+
+            return ExifHandlerPiexif
+
+        if return_value is not None:
             return return_value
 
-        if piexif is None:
-            return stub
-        return func
+        class NoExifHandler:
```

Comment:
As mentioned above, I think it would be nice to define the class outside of the decorator to avoid the nesting.